File: tests/test_loop.py
"""End-to-end loop tests: echo harnesses, real git in tmp repos only."""
import json
import sys
import tempfile
import unittest
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from src.orchestrator import Orchestrator  # noqa: E402
from src import statemachine  # noqa: E402

import helpers  # noqa: E402


class LoopTest(unittest.TestCase):
    def setUp(self):
        self.tmp = tempfile.TemporaryDirectory()
        self.addCleanup(self.tmp.cleanup)
        self.dir = Path(self.tmp.name)
        self.repo = helpers.make_git_repo(self.dir / "repo")
        self.config = helpers.write_echo_config(self.dir / "echo.toml")
        self.logs = []

    def _orch(self, mission_path, **kw):
        kw.setdefault("auto", True)
        kw.setdefault("log", self.logs.append)
        return Orchestrator(tool_dir=helpers.TOOL_DIR,
                            mission_path=mission_path,
                            config_path=self.config, **kw)

    def test_full_loop_echo_reaches_ready(self):
        mission = helpers.write_mission(self.dir / "m.md", self.repo)
        orch = self._orch(mission)
        rc = orch.run()
        self.assertEqual(rc, 0)
        state = statemachine.load(orch.run_dir)
        self.assertEqual(state.phase, "READY")
        # The echo lane file was delivered into the target branch.
        delivered = self.repo / "src" / "example" / "echo-L1-w0.md"
        self.assertTrue(delivered.is_file())
        # Worktrees and gauntlet branches were cleaned up.
        self.assertFalse(Path(state.worktrees[0]).exists())
        branches = helpers.git(self.repo, "branch", "--list", "gauntlet/*")
        self.assertEqual(branches.strip(), "")
        # Run directory has the expected artifacts.
        run_dir = Path(state.run_dir)
        self.assertTrue((run_dir / "mission.md").is_file())
        self.assertTrue((run_dir / "config.toml").is_file())
        self.assertTrue((run_dir / "state.json").is_file())
        self.assertTrue((run_dir / "report.md").is_file())
        review = json.loads(
            (run_dir / "verdicts" / "review-w0.json").read_text())
        self.assertEqual(review, {"groups": []})
        # Lane capsule was deleted after successful integration.
        self.assertFalse(
            (run_dir / "capsules" / "implementer-L1-w0.md").exists())

    def test_full_loop_without_lanes_uses_planner(self):
        mission = helpers.write_mission(self.dir / "m.md", self.repo,
                                        lanes=[])
        orch = self._orch(mission)
        rc = orch.run()
        self.assertEqual(rc, 0)
        state = statemachine.load(orch.run_dir)
        self.assertEqual(state.phase, "READY")
        # Echo planner cut lane E1 owning "**"; echo wrote at repo root.
        self.assertTrue((self.repo / "echo-E1-w0.md").is_file())
        self.assertEqual([l.id for l in state.lanes], ["E1"])

    def test_dry_run_prints_commands_and_changes_nothing(self):
        mission = helpers.write_mission(self.dir / "m.md", self.repo)
        orch = self._orch(mission, dry_run=True)
        rc = orch.run()
        self.assertEqual(rc, 0)
        state = statemachine.load(orch.run_dir)
        # Nothing executed -> nothing integrated -> READY_NO_CHANGE.
        self.assertEqual(state.phase, "READY_NO_CHANGE")
        out = "\n".join(self.logs)
        self.assertIn("DRY-RUN: (cd", out)
        self.assertIn("git worktree add", out)
        self.assertIn("git branch -D", out)  # cleanup commands printed
        self.assertIn("&& true)", out)  # gate command printed, not executed
        # The repo was not modified (only the run record under .missions/).
        self.assertFalse((self.repo / "src").exists())
        status = helpers.git(self.repo, "status", "--porcelain")
        self.assertEqual([l for l in status.splitlines()
                          if ".missions" not in l], [])

    def test_resume_after_init_completes(self):
        mission = helpers.write_mission(self.dir / "m.md", self.repo)
        orch = self._orch(mission)
        orch._phase_init()  # stop right after INIT
        self.assertEqual(orch.state.phase, "PLAN")
        run_dir = orch.run_dir

        resumed_logs = []
        orch2 = Orchestrator(tool_dir=helpers.TOOL_DIR, resume_dir=run_dir,
                             auto=True, log=resumed_logs.append)
        rc = orch2.run()
        self.assertEqual(rc, 0)
        self.assertEqual(orch2.state.phase, "READY")
        self.assertEqual(orch2.state.run_id, orch.state.run_id)
        self.assertTrue(
            (self.repo / "src" / "example" / "echo-L1-w0.md").is_file())

    def test_init_refuses_staged_changes(self):
        (self.repo / "dirty.txt").write_text("staged\n", encoding="utf-8")
        helpers.git(self.repo, "add", "dirty.txt")
        mission = helpers.write_mission(self.dir / "m.md", self.repo)
        orch = self._orch(mission)
        rc = orch.run()
        self.assertEqual(rc, 2)
        self.assertEqual(orch.state.phase, "BLOCKED")
        self.assertIn("staged changes", orch.state.blocked_reason)

    def test_overlapping_prewritten_lanes_blocked(self):
        mission = helpers.write_mission(
            self.dir / "m.md", self.repo, lanes=[
                {"lid": "L1", "owns": '"src/**"'},
                {"lid": "L2", "owns": '"src/example/**"'},
            ])
        orch = self._orch(mission)
        rc = orch.run()
        self.assertEqual(rc, 2)
        self.assertEqual(orch.state.phase, "BLOCKED")
        self.assertIn("overlap", orch.state.blocked_reason)


if __name__ == "__main__":
    unittest.main()

File: rust/tests/test_loop.rs
use std::path::{Path, PathBuf};
use std::sync::{Arc, Mutex};
use tempfile::tempdir;

use gauntlet::orchestrator::Orchestrator;
use gauntlet::statemachine::load;

mod helpers;
use helpers::{git, make_git_repo, write_echo_config, write_mission, LaneSpec};

#[test]
fn test_full_loop_echo_reaches_ready() {
    let tmp = tempdir().unwrap();
    let repo = make_git_repo(&tmp.path().join("repo"));
    let config = write_echo_config(&tmp.path().join("echo.toml"));
    let mission = write_mission(&tmp.path().join("m.md"), &repo, "example", None, None);

    let logs = Arc::new(Mutex::new(Vec::new()));
    let logs_clone = Arc::clone(&logs);
    let log_fn = Arc::new(move |msg: &str| {
        logs_clone.lock().unwrap().push(msg.to_string());
    });

    let tool_dir = PathBuf::from(".");
    let mut orch = Orchestrator::new(
        &tool_dir,
        Some(&mission),
        None,
        Some(&config),
        true,
        false,
        None,
        false,
        0,
        2,
        Some(log_fn),
    )
    .unwrap();

    let rc = orch.run();
    assert_eq!(rc, 0);

    let state = load(orch.run_dir.as_ref().unwrap()).unwrap();
    assert_eq!(state.phase, "READY");

    // The echo lane file was delivered into target branch
    let delivered = repo.join("src").join("example").join("echo-L1-w0.md");
    assert!(delivered.is_file(), "delivered file should exist at {}", delivered.display());

    // Worktrees and gauntlet branches were cleaned up
    assert!(!Path::new(&state.worktrees[0]).exists());
    let branches = git(&repo, &["branch", "--list", "gauntlet/*"]);
    assert_eq!(branches.trim(), "");

    // Run dir artifacts exist
    let run_dir = Path::new(state.run_dir.as_ref().unwrap());
    assert!(run_dir.join("mission.md").is_file());
    assert!(run_dir.join("config.toml").is_file());
    assert!(run_dir.join("state.json").is_file());
    assert!(run_dir.join("report.md").is_file());

    let review_text = std::fs::read_to_string(run_dir.join("verdicts").join("review-w0.json")).unwrap();
    let review_val: serde_json::Value = serde_json::from_str(&review_text).unwrap();
    assert_eq!(review_val, serde_json::json!({ "groups": [] }));

    // Lane capsule was deleted after successful integration
    assert!(!run_dir.join("capsules").join("implementer-L1-w0.md").exists());
}

#[test]
fn test_full_loop_without_lanes_uses_planner() {
    let tmp = tempdir().unwrap();
    let repo = make_git_repo(&tmp.path().join("repo"));
    let config = write_echo_config(&tmp.path().join("echo.toml"));
    let mission = write_mission(&tmp.path().join("m.md"), &repo, "example", Some(&[]), None);

    let tool_dir = PathBuf::from(".");
    let mut orch = Orchestrator::new(
        &tool_dir,
        Some(&mission),
        None,
        Some(&config),
        true,
        false,
        None,
        false,
        0,
        2,
        None,
    )
    .unwrap();

    let rc = orch.run();
    assert_eq!(rc, 0);

    let state = load(orch.run_dir.as_ref().unwrap()).unwrap();
    assert_eq!(state.phase, "READY");
    assert!(repo.join("echo-E1-w0.md").is_file());
    assert_eq!(state.lanes.iter().map(|l| l.id.as_str()).collect::<Vec<_>>(), vec!["E1"]);
}

#[test]
fn test_dry_run_prints_commands_and_changes_nothing() {
    let tmp = tempdir().unwrap();
    let repo = make_git_repo(&tmp.path().join("repo"));
    let config = write_echo_config(&tmp.path().join("echo.toml"));
    let mission = write_mission(&tmp.path().join("m.md"), &repo, "example", None, None);

    let logs = Arc::new(Mutex::new(Vec::new()));
    let logs_clone = Arc::clone(&logs);
    let log_fn = Arc::new(move |msg: &str| {
        logs_clone.lock().unwrap().push(msg.to_string());
    });

    let tool_dir = PathBuf::from(".");
    let mut orch = Orchestrator::new(
        &tool_dir,
        Some(&mission),
        None,
        Some(&config),
        true,
        true,
        None,
        false,
        0,
        2,
        Some(log_fn),
    )
    .unwrap();

    let rc = orch.run();
    assert_eq!(rc, 0);

    let state = load(orch.run_dir.as_ref().unwrap()).unwrap();
    assert_eq!(state.phase, "READY_NO_CHANGE");

    let out = logs.lock().unwrap().join("\n");
    assert!(out.contains("DRY-RUN: (cd"));
    assert!(out.contains("git worktree add"));
    assert!(out.contains("git branch -D"));
    assert!(out.contains("&& true)"));

    assert!(!repo.join("src").exists());
}

#[test]
fn test_resume_after_init_completes() {
    let tmp = tempdir().unwrap();
    let repo = make_git_repo(&tmp.path().join("repo"));
    let config = write_echo_config(&tmp.path().join("echo.toml"));
    let mission = write_mission(&tmp.path().join("m.md"), &repo, "example", None, None);

    let tool_dir = PathBuf::from(".");
    let mut orch = Orchestrator::new(
        &tool_dir,
        Some(&mission),
        None,
        Some(&config),
        true,
        false,
        None,
        false,
        0,
        2,
        None,
    )
    .unwrap();

    orch._phase_init().unwrap();
    assert_eq!(orch.state.phase, "PLAN");
    let run_dir = orch.run_dir.clone().unwrap();

    let mut orch2 = Orchestrator::new(
        &tool_dir,
        None,
        Some(&run_dir),
        None,
        true,
        false,
        None,
        false,
        0,
        2,
        None,
    )
    .unwrap();

    let rc = orch2.run();
    assert_eq!(rc, 0);
    assert_eq!(orch2.state.phase, "READY");
    assert_eq!(orch2.state.run_id, orch.state.run_id);
    assert!(repo.join("src").join("example").join("echo-L1-w0.md").is_file());
}

#[test]
fn test_init_refuses_staged_changes() {
    let tmp = tempdir().unwrap();
    let repo = make_git_repo(&tmp.path().join("repo"));
    std::fs::write(repo.join("dirty.txt"), "staged\n").unwrap();
    git(&repo, &["add", "dirty.txt"]);

    let config = write_echo_config(&tmp.path().join("echo.toml"));
    let mission = write_mission(&tmp.path().join("m.md"), &repo, "example", None, None);

    let tool_dir = PathBuf::from(".");
    let mut orch = Orchestrator::new(
        &tool_dir,
        Some(&mission),
        None,
        Some(&config),
        true,
        false,
        None,
        false,
        0,
        2,
        None,
    )
    .unwrap();

    let rc = orch.run();
    assert_eq!(rc, 2);
    assert_eq!(orch.state.phase, "BLOCKED");
    assert!(orch.state.blocked_reason.unwrap().contains("staged changes"));
}

#[test]
fn test_overlapping_prewritten_lanes_blocked() {
    let tmp = tempdir().unwrap();
    let repo = make_git_repo(&tmp.path().join("repo"));
    let config = write_echo_config(&tmp.path().join("echo.toml"));
    let lanes = vec![
        LaneSpec { lid: "L1", owns: "\"src/**\"", forbidden: "" },
        LaneSpec { lid: "L2", owns: "\"src/example/**\"", forbidden: "" },
    ];
    let mission = write_mission(&tmp.path().join("m.md"), &repo, "example", Some(&lanes), None);

    let tool_dir = PathBuf::from(".");
    let mut orch = Orchestrator::new(
        &tool_dir,
        Some(&mission),
        None,
        Some(&config),
        true,
        false,
        None,
        false,
        0,
        2,
        None,
    )
    .unwrap();

    let rc = orch.run();
    assert_eq!(rc, 2);
    assert_eq!(orch.state.phase, "BLOCKED");
    assert!(orch.state.blocked_reason.unwrap().contains("overlap"));
}
